### PR TITLE
Metropolis Algorithm for MCMC

### LINKED ISSUE #1260: [TASK] Markov Chain Monte Carlo: Metropolis Algorithm
--------
Issue Description
--------
Markov Chain Monte Carlo (MCMC) is mainly used for Bayesian Calibration (or inverse uncertainty quantification). It can also be used to sample any given distribution. Currently, this capability is not available in RAVEN. We would like to implement several MCMC algorithms which will be part of the Bayesian calibration implementation within RAVEN framework.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #1260
This PR will also update the submodule ID of LOGOS. Recent changes in RAVEN dependencies file structure caused the failure of LOGOS regression test. This has been recently fixed in LOGOS. see issue #1114  
In addition, this PR also make some improvement on the restart capability (see #1280), this is because metadata is required for several Samplers when restarted is requested. 
 
Implementation of Markov Chain Monte Carlo algorithm, i.e. Metropolis Algorithm.

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Samplers/MCMC/MCMC.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    self._likelihood = None # stores the output from the likelihood
+    self._availProposal = {'normal': Distributions.Normal(0.0, 1.0),
+                           'uniform': Distributions.Uniform(-1.0, 1.0)} # available proposal distributions
+    self._acceptDist = Distributions.Uniform(0.0, 1.0) # uniform distribution for accept/rejection purpose
+    # assembler objects
+    self.addAssemblerObject('proposal', '-n', True)
+
+  def localInputAndChecks(self, xmlNode, paramInput):
+    """
+      unfortunately-named method that serves as a pass-through for input reading.
```
Comment (alfoa) [inline framework/Samplers/MCMC/MCMC.py:139]: unfortunately-named method? O_o

**Comment by alfoa** [comment]
@wangcj05 I removed most of the comments I had because I need a clarification... I do not understand the <proposal> <distribution> link in your sampler... the proposal is the actual tool for "sampling" and the <distribution> is for the ***SampledVarsPb***? Why we have this need? Isn't it redundant? I also check the algorithm and still I do not get the need to have both "distributions".

File: framework/Samplers/MCMC/MCMC.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply framework/Samplers/MCMC/MCMC.py:139]: 10:1 says I wrote the original version this was copied from ... heh

File: framework/Samplers/MCMC/MCMC.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        self._initialValues[var] = initNode.value
+      else:
+        self._initialValues[var] = None
+      proposal = varNode.findFirst('proposal')
+      if proposal:
+        self._proposal[var] = proposal.value
+      else:
+        self._proposal[var] = None
+    # TargetEvaluation Node (Required)
+    targetEval = paramInput.findFirst('TargetEvaluation')
```
Comment (PaulTalbot-INL) [inline framework/Samplers/MCMC/MCMC.py:192]: I believe the AdaptiveSampler does this for you, so it should not be needed here if you call the base class, right?

**Comment by alfoa** [comment]
@wangcj05  I like the generality of allowing the user to use any distribution for the proposal (following the general metropolis-hasting algo), but I am afraid that we will "pay" this generality in terms of convergence issues and/or unexpected behavior. For example, how do we handle the discrete distributions here? Multivariates?

**Comment by alfoa** [comment]
@wangcj05 as during our conversation, if we limit the usability of the proposal distributions and check the truncability, I am good for the merge

File: framework/Distributions.py [hunk 3]
```diff
@@ -439,6 +439,15 @@ def pdf(self,x):
     returnPdf = self._distribution.pdf(x)
     return returnPdf
 
+  def logpdf(self,x):
```
Comment (mandd) [inline framework/Distributions.py:442]: I am having my revenge here: should it be logPdf?

File: framework/Samplers/MCMC/MCMC.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  @classmethod
+  def userManualDescription(cls):
+    """
+      Provides a user manual description for this actor. Should only be needed for base classes.
+      @ In, None
+      @ Out, descr, string, description
+    """
+    descr = r"""
+    \section{Markov Chain Monte Carlo} \label{sec:MCMC}
+    The Markov chain Monte Carlo (MCMC) is another important entity in the RAVEN framework.
```
Comment (mandd) [inline framework/Samplers/MCMC/MCMC.py:47]: ... is another important sampler ...

File: framework/Samplers/MCMC/MCMC.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    for key, value in self._updateValues.items():
+      self.values[key] = value
+      self.inputInfo['SampledVarsPb'][key] = self.distDict[key].pdf(value)
+      self.inputInfo['ProbabilityWeight-' + key] = 1.
+    self.inputInfo['PointProbability'] = 1.0
+    self.inputInfo['ProbabilityWeight' ] = 1.0
+
+  def localFinalizeActualSampling(self, jobObject, model, myInput):
+    """
+      General function (available to all samplers) that finalize the sampling
```
Comment (mandd) [inline framework/Samplers/MCMC/MCMC.py:247]: finalizes

File: framework/Samplers/MCMC/MCMC.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    self._targetEvaluation = targetEval.value
+    self._updateValues = copy.copy(self._initialValues)
+
+  def initialize(self, externalSeeding=None, solutionExport=None):
+    """
+      This function should be called every time a clean MCMC is needed. Called before takeAstep in <Step>
+      @ In, externalSeeding, int, optional, external seed
+      @ In, solutionExport, DataObject, optional, a PointSet to hold the solution
+      @ Out, None
+    """
```
Comment (mandd) [inline framework/Samplers/MCMC/MCMC.py:202]: This is a generic question: does this MCMC method allow the user to use a chain generated in the past that needs to be augmented?

**Comment by mandd** [comment]
I have completed a first round of comments; I need a bit more time to think more about it

File: framework/Samplers/MCMC/MCMC.py (hunk 1 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py]: fixed

File: framework/Distributions.py (hunk 3 above)
Comment (wangcj05) [reply framework/Distributions.py]: fixed.

File: framework/Samplers/MCMC/MCMC.py (hunk 4 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py]: fixed

File: framework/Samplers/MCMC/MCMC.py (hunk 5 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py]: fixed

File: framework/Samplers/MCMC/MCMC.py (hunk 6 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py:206]: I think the answer for this question is No for current implementation. This seems to me similar to Restart option.

**Comment by wangcj05** [comment]
@alfoa I have added the checks for Distributions. Only continuous distribution is allowed.

**Comment by wangcj05** [comment]
@mandd I have addressed your comments.

File: framework/Samplers/MCMC/MCMC.py (hunk 6 above)
Comment (alfoa) [reply framework/Samplers/MCMC/MCMC.py:206]: No restart?

Comment (alfoa) [reply framework/Samplers/MCMC/MCMC.py:206]: Can you try a test for it?

File: tests/framework/MCMC/test_metropolis.xml [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    <author>wangc</author>
+    <created>2020-06-26</created>
+    <classesTested>MCMC.Metropolis</classesTested>
+    <description>
+      Test the Markov Chain Monte Carlo algorithm: Metropolis Sampling
+      A 2-D multivariate normal distribution is used as the likelihood function.
+      mean: [5, 5], cov=[[1, 0.9], [0.9, 1]]
+      Both input parameters have the standard normal distribution as their prior distribution.
+      The proposal distributions for the input variables are also standard normal distribution.
+      "TargetEvaluation" is used to collect the inputs and outputs from the likelihood model.
```
Comment (alfoa) [inline tests/framework/MCMC/test_metropolis.xml:14]: since this goes into a latex file, I would use the ``  '' construct for the quotes

Comment (wangcj05) [reply tests/framework/MCMC/test_metropolis.xml]: fixed

**Comment by wangcj05** [comment]
@alfoa I have opened an issue for the restart problem, see #1280. As we discussed, it is a basic issue for raven, and we need a discussion on this. I think a different PR need to be created to address the issue. Please let me know what you think.

**Comment by alfoa** [comment]
@wangcj05  do you want to develop the restart here or in another PR?

**Comment by wangcj05** [comment]
> @wangcj05 do you want to develop the restart here or in another PR?

I have updated the Metropolis Sampler with the restart capability. The original test has also been updated with the restart.

**Comment by wangcj05** [comment]
@alfoa I have addressed all your comments, restart is now tested, and the tests are green now.

**Comment by alfoa** [comment]
Checklist passed...

Comments have been addressed...

I approve this PR...

Merging...

@wangcj05 can you update the issue #1225 for the new release changelog?

**Pull request merged**